src/core/src/handle.rs

```diff
@@ -16,20 +16,82 @@
 
 //! Device resource handles
 
-use std::{cmp, hash, mem};
+use std::{cmp, hash, mem, ops};
 use std::marker::PhantomData;
-use std::sync::Arc;
-use {shade, tex, Resources};
+use std::sync::{Arc, Weak, Mutex, MutexGuard};
+use {mapping, shade, tex, Resources};
 use factory::{BufferInfo, Typed};
 
 
 /// Raw (untyped) Buffer Handle
-#[derive(Clone, Debug, Eq, Hash, PartialEq)]
-pub struct RawBuffer<R: Resources>(Arc<R::Buffer>, BufferInfo);
+#[derive(Clone, Debug)]
+pub struct RawBuffer<R: Resources>(Arc<RawBufferData<R>>);
+
+/// Raw buffer handle data
+#[derive(Debug)]
+#[allow(missing_docs)]
+pub struct RawBufferData<R: Resources> {
+    pub resource: R::Buffer,
+    pub info: BufferInfo,
+    pub mapping: Mutex<Option<Weak<mapping::Raw<R>>>>
+}
 
 impl<R: Resources> RawBuffer<R> {
+    /// Get raw buffer resource
+    pub fn resource(&self) -> &R::Buffer { &(self.0).resource }
+
     /// Get raw buffer info
-    pub fn get_info(&self) -> &BufferInfo { &self.1 }
+    pub fn get_info(&self) -> &BufferInfo { &(self.0).info }
+
+    /// Is the requested mapping access matching the expected usage ?
+    pub fn valid_access(&self, access: mapping::Access) -> Result<(), mapping::Error> {
+        self.get_info().usage.valid_access(access)
+    }
+
+    fn mapping_opt(&self) -> MutexGuard<Option<Weak<mapping::Raw<R>>>> {
+        (self.0).mapping.lock().unwrap()
+    }
+
+    /// Get current mapping
+    pub fn mapping(&self) -> Option<RawMapping<R>> {
+        let weak_opt = self.mapping_opt();
+        weak_opt.as_ref().map(|w| RawMapping(w.upgrade().unwrap()))
+    }
+
+    /// Needs to be called internally after the buffer is mapped
+    fn was_mapped(&self, raw: &RawMapping<R>) {
+        let mut weak_opt = self.mapping_opt();
+        if weak_opt.is_some() { panic!("attempted to map a buffer twice"); }
+        *weak_opt = Some(Arc::downgrade(&raw.0));
+    }
+
+    /// Needs to be called internally after the buffer is unmapped
+    pub fn was_unmapped(&self) {
+        let mut weak_opt = self.mapping_opt();
+        *weak_opt = None;
+    }
```

Comment:
For the persistent mapping? Hmm... best of all would be to defer that decision to the backend itself, if possible. Otherwise, let's just allow it for now.